[서브웨이] 3주차 -로또 코드리뷰 요청입니다.

joseph415 opened the pull request.

joseph415 committed:
[docs] README에 요구사항, 기능 구현 목록 작성

joseph415 committed:
[feat] 1000원 단위가 아닌 경우 재입력 기능 구현

joseph415 committed:
[feat] 숫자 외의 값이 들어오면 예외처리 기능 구현

joseph415 committed:
[feat] 음수인 경우예외처리 기능 구현

joseph415 committed:
[feat] 천원단위로 로또 개수 반환하는 기능 구현

joseph415 committed:
[feat] 남은 돈 반환 기능 구현

joseph415 committed:
[feat] 1000원 미만 입력시 거스름돈 반환 후 재입력 기능 구현

joseph415 committed:
[feat] lottoBall 번호 값 검증 기능 구현

joseph415 committed:
[feat] 개수에 맞춰 로또 자동으로 발급하는 기능 구현

joseph415 committed:
[feat] 당첨번호 입력받기, 보너스 번호 입력받기 기능 구현

joseph415 committed:
[feat] 당첨된 로또 번호 개수 세는 기능 구현

joseph415 committed:
[feat] 당첨번호와 로또 티켓을 비교하는 기능 구현

joseph415 committed:
[feat] 당첨 결과 enum 생성

joseph415 committed:
[feat] 당첨 등수 구하는 기능 구현

joseph415 committed:
[refactor] LottoTicket에 null 값 들어올 경우에 대한 예외처리 구현

joseph415 committed:
[refactor] customException 생성

joseph415 committed:
[refactor] LottoController 리팩토링

joseph415 committed:
[feat] 통계 구하는 로직 구현

joseph415 committed:
[refactor] 전반적인 Convention 리팩토링

joseph415 committed:
[refactor] LottoBallFactory 리팩토링

joseph415 committed:
[test] 원하는 로또값이 나오는지 확인 테스트

joseph415 committed:
[test] 로또 볼이 비어있을 경우 테스트

joseph415 committed:
[test] 로또 티켓이 재대로 생성되었을 경우 테스트

joseph415 committed:
[refactor] 구입금액 테스트 리팩토링

joseph415 committed:
[test] 구입금액이 천원 미만의 값이 들어 왔을때 테스트

joseph415 committed:
[refactor] 보너스 번호가 있는 경우 테스트

joseph415 committed:
[refactor] 1등부터 꼴등까지 골라내는 테스트

joseph415 committed:
[refactor] 전반적인 리팩토링

joseph415 committed:
[feat] 2등일 경우 print방식 하는 기능 구현

joseph415 committed:
[refactor] view와 domain분리

joseph415 committed:
[refactor] 당첨 결과를 print 하는 메소드 리팩토링

joseph415 committed:
[refactor] OutputView stream 리팩토링

joseph415 committed:
[refactor] 메소드 명명과 WinningRank  리팩토링

joseph415 committed:
[feat] Lottoticket발생시 로또 번호 개수가 7개 이상일경우 예외발생하는 로직과 수익율 계산시 예외 추가 구현

joseph415 committed:
[refactor] WinningRank 에 boolean 인자 추가하고 second thrid Rank 비교 메소드 삭제

dave915 commented on src/test/java/lotto/domain/WinningRankTest.java:
```diff
@@ -0,0 +1,39 @@
+
+    @ParameterizedTest
+    @DisplayName("1등부터 꼴등까지 골라내는 테스트")
+    @CsvSource(value = {"1,2,3,4,5,6:FIRST_RANK",
+            "1,2,3,4,5,7:SECOND_RANK",
+            "1,2,3,4,5,10:THIRD_RANK",
+            "1,2,3,4,44,45:FOURTH_RANK",
+            "1,2,3,43,44,45:FIFTH_RANK",
+            "1,2,40,41,42,43:NO_RANK"}, delimiter = ':')
```
2번째와 4번째 테스트가 실패하네요 😭

joseph415 replied:
리팩토링하면서 enum상수값을 잘못 바꿔주었네요,, 테스트를 돌려본줄 알았는데 죄송합니다 ㅠ 
다음에는 다시한번확인하겠습니다.!

dave915 commented on src/test/java/lotto/domain/WinningRankTest.java:
```diff
@@ -0,0 +1,39 @@
+        WinningBalls winningBalls = new WinningBalls(winningBallValues, bonus);
+        int correctCount = winningBalls.hitLottoBalls(new LottoTicket(lottoTicket));
+        boolean correctBonusNumber = winningBalls.hitBonusBall(new LottoTicket(lottoTicket));
+        Assertions.assertThat(winningRank).isEqualTo(WinningRank.determineRank(correctCount, correctBonusNumber));
```
예상값을 뒤에 넣어주셔야 Junit테스트 실패시 
Expected가 winningRank값
Actual이 WinningRank.determineRank(correctCount, correctBonusNumber) 값이 나옵니다 :)
```
Assertions.assertThat(WinningRank.determineRank(correctCount, correctBonusNumber)).isEqualTo(winningRank);
```

joseph415 replied:
아 사소한것에 실수가 있었네요 assertj 다시 한번 공부하겠습니다 :)

dave915 commented on src/main/java/lotto/domain/WinningRank.java:
```diff
@@ -0,0 +1,43 @@
+
+    public static WinningRank determineRank(int correctNumber, boolean isBonusNumber) {
+        return Arrays.stream(values())
+                .filter(result -> result.winningBallCount == correctNumber && result.isBonusBall == isBonusNumber)
```
filter 내부가 명시적으로 보기 힘들어요 😭
메소드 분리해 어떤 판별을 하는건지 명시적으로 보이게 하면 좋을 것 같아요.

dave915 commented on src/main/java/lotto/domain/WinningBalls.java:
```diff
@@ -0,0 +1,35 @@
+import java.util.List;
+
+public class WinningBalls {
+    private final List<LottoBall> winningBalls;
```
LottoTicket과 winningBalls의 다른점은 무엇일까요? 🤔

joseph415 replied:
LottoTicket은  티켓을 만들어 주는 역할을 위해 만들었고
WinningBalls는 지난주 당첨번호티켓를 담기위해 만든 클래스였습니다..혹시 두 클래스를 합처야 하나요?
우선은 의미를 명확하게 하기 위해 WinningTicket 이라고 클래스명을 바꾸겠습니다!

joseph415 replied:
혹시 일급컬랙션에 인스턴스 변수가 final로 선언이 안되었던것이 다른점이였나요? 수정했습니다!
-> 본질적으로 같다는 의미

dave915 commented on src/main/java/lotto/domain/WinningBalls.java:
```diff
@@ -0,0 +1,35 @@
+
+    private void validateWinningBallsWithDuplicatedBonusBall() {
+        if (winningBalls.contains(this.BonusBall)) {
+            throw new DuplicationException("보너스 볼이 중복입니다. 당첨 번호를 다시 입력해주세요.");
```
CustomException으로 추가적으로 해주는일이 없다면 표준 예외를 사용하는 것이 어떨까요? 

아래 글 중 Item 72. 표준 예외를 사용하라를 읽어보세요!
https://badcandy.github.io/2019/01/01/effective-java-3rd-chapter10/

dave915 commented on src/main/java/lotto/domain/LottoBallFactory.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    static {
+        lottoBalls = IntStream.rangeClosed(1, 45)
```
매직넘버 사용은 피해주세요! 

https://hoonmaro.tistory.com/44

dave915 commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,74 @@
+
+    private void generateLottoTickets() {
+        for (int i = 0; i < purchaseAmount.lottoTicket(); i++) {
+            Collections.shuffle(LottoBallFactory.getInstance());
```
로또번호를 매번 shuffle 하는 책임은 LottoBallFactory가 갖고 있어야 하지 않을까요? 
서브웨이 생각은 어떠신가요?

joseph415 replied:
로또 볼을 만들고 그 볼을 관리 하는 factoty클래스에서 볼에 대한 책임을 갖는것이 맞는것 같습니다.
감사합니다 :)

dave915 commented on src/main/java/lotto/domain/LottoBallFactory.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    public static List<LottoBall> getInstance() {
+        return lottoBalls;
```
lottoBalls 참조값을 넘겨주면 외부에서 원하는 범위만큼 가져가서 조작 할 수 있지 않을까요?

joseph415 replied:
이 의미를 잘 이해하지 못했는데 파라미터로 원하는 범위를 넣어주면 그만큼 가져가서 쓰도록 하라는 말인가요??

dave915 commented on src/main/java/lotto/domain/PurchaseAmount.java:
```diff
@@ -0,0 +1,34 @@
+
+    private void validateLottoUnit() {
+        if (this.purchaseAmount < LOTTO_PURCHASE_UNIT) {
+            throw new NotBuyLottoTicketException("한개도 구매할 수 없습니다." + this.purchaseAmount + "원을 반환합니다.");
```
String.format을 사용해보세요 :)

joseph415 replied:
좋은 api기능이 있었네요! 수정했습니다 :)

dave915 commented on src/main/java/lotto/util/InputValidationUtil.java:
```diff
@@ -0,0 +1,25 @@
+        try {
+            return Integer.parseInt(inputValue);
+        } catch (NumberFormatException e) {
+            throw new NumberFormatException("숫자가 아닙니다. 재입력 해주세요.");
```
에러메세지도 상수로 올리면 어떨까요?

dave915 commented on src/main/java/lotto/util/WinningBallsUtils.java:
```diff
@@ -0,0 +1,48 @@
+    private List<LottoBall> winningBalls;
+
+    public WinningBallsUtils(String input) {
+        String[] winningBalls = input.split(",");
```
하드코딩은 피해주세요! `","`
그리고 관련된 코드별로 공백을주면 가독성이 더 좋아집니다 :) 

```
    public WinningBallsUtils(String input) {
        String[] winningBalls = input.split(",");
        
        validateWinningBallsNumber(winningBalls);
        validateWinningBallsLength(winningBalls);
        validateDuplicatedWinningBalls(winningBalls);
        
        this.winningBalls = Arrays.stream(winningBalls)
                .map(Integer::parseInt)
                .map(LottoBallFactory::findByLottoBall)
                .collect(Collectors.toList());
        Collections.sort(this.winningBalls);
    }
```

dave915 commented on src/main/java/lotto/util/WinningBallsUtils.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    private void validateDuplicatedWinningBalls(String[] winningBalls) {
+        Set<String> DuplicatedValidation = new HashSet<>(Arrays.asList(winningBalls));
```
변수명의 첫글자는 소문자로 해주세요 :)

dave915 commented on src/main/java/lotto/util/WinningBallsUtils.java:
```diff
@@ -0,0 +1,48 @@
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class WinningBallsUtils {
```
이 클래스는 유틸이라기 보단 winningBalls의 도메인 로직이라고 보이네요 🤔

joseph415 replied:
도메인로직이 있는 부분을 따로 WinningTicket으로 분리하였습니다.

dave915 commented on src/main/java/lotto/domain/EarningRate.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    public int countRankPeople(List<WinningRank> winningRanks, WinningRank winningRank) {
+        return (int) winningRanks.stream().filter(Rank -> winningRank == Rank).count();
```
람다식에서 파라미터 부분은 변수이기 때문에 앞글자를 소문자로 해주세요. 
`filter(rank -> winningRank == rank)`

joseph415 replied:
이전 미션부터 사소한 것에서 실수가 많이 나오네요,, 이런 실수를 줄이는게 가장중요한것 같네요ㅠㅠ 꼭 다음 번에는 꼼꼼히 확인하겠습니다!!😀

dave915 reviewed (changes requested):
서브웨이 안녕하세요! 리뷰를 맡게된 데이브 입니다. 
구현 잘 하셨네요 😀
몇가지 피드백 남겼으니 확인 후 반영 부탁드릴게요! 
추가로 궁금한 부분이 있다면 편하게 DM주세요!

joseph415 committed:
[refacor] 리뷰 refactoring

WinningRank 테스트의 예상값과 실제값의 인자 위치를 변경
WinningRank의 determineRank 메소드의 filter 내부를 명시적으로 표현하기위해 메소드로 분리
LottoTicket과 WinningBalls 클래스의 의미를 명확하게 하기 위해 WinningTicket 이라고 클래스명을 변경
DuplicationException을 표준예외로 변경
매직넘버 상수화 리팩토링후 테스트 확인
String Format으로 사용
WinningBallUtils 가 인스턴스 변수를 갖고 비지니스 로직을 갖는 부분이 있어서 WinningTicketFactory 클래스를 만들어 검증유틸 부분과 비지니스 로직을 나눴고 테스트를 추가
에러메시지 상수화
전체적인 리팩토링

joseph415 committed:
[refactor] 리뷰받은 부분 refactoring

custom exception을 모두 표준예외로 변경
LottoBallFactory에서 인스턴스 참조값을 주었던 메소드가 외부에서 변경가능성이 있어서 필요한 부분만 unmidifiableList를 통해 리턴해주도록 변경
LottoTicket과 WinningBallFactory의 본질적 기능이 같아 합침

dave915 reviewed (approved):
피드백 반영 한 부분 확인 했어요 💯
다음 Step 진행을 위해 여기서 머지 할게요! 
혹시나 궁금한 점이 있으시면 언제든지 질문주세요!
